Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp

```diff
@@ -733,4 +734,50 @@ void TreeBuilder::missing_cells_fixup(Vector<JS::Handle<Box>> const& table_root_
         });
     }
 }
+
+static bool is_row_group(Node const& node)
+{
+    // NB: For the purpose of reordering row groups, rows can be treated as a group with one element.
+    return node.display().is_table_header_group() || node.display().is_table_row_group() || node.display().is_table_footer_group() || node.display().is_table_row();
+}
+
+void TreeBuilder::reorder_row_groups(NodeWithStyle& root)
+{
+    for_each_in_tree_with_inside_display<CSS::Display::Inside::Table>(root, [&](auto& parent) {
+        Vector<JS::Handle<Node>> row_groups;
+        // Per https://www.w3.org/TR/css-tables-3/#table-header-group:
+        // Like table-row-group but, for layout purposes, the first such row group is always displayed before all other rows and row groups.
+        const Node* header_group { nullptr };
+        for (auto child = parent.first_child(); child; child = child->next_sibling())
+            if (child->display().is_table_header_group()) {
+                row_groups.append(child);
+                header_group = child;
+                break;
+            }
+        JS::Handle<Node> footer_group;
+        for (auto child = parent.first_child(); child; child = child->next_sibling()) {
+            if (!is_row_group(*child))
+                continue;
+            if (child->display().is_table_footer_group() && !footer_group)
+                footer_group = child;
+            // Accumulate table row groups and header / footer groups which aren't the first and behave like regular row groups.
+            else if (child != header_group)
+                row_groups.append(child);
+        }
```

Comment:
it looks like we can collect header, footer and the rest of table groups in a single loop.

Reply:
We also have to end up with the row groups in the correct order. I'm not sure how you'd do that cleanly with one loop.

Reply:
hmmm, why would pushing a row group into `row_groups` as you iterating children won't give you desired order? as I understand it should preserve order of everything besides header and footer nodes so I don't see why it is not possible to do in one go.

Reply:
Feel free to take over.